pkg/ingester/ingester_test.go

```diff
@@ -503,190 +505,245 @@ func TestIngesterAppendBlankLabel(t *testing.T) {
-		require.NoError(t, err)
-		assert.Equal(t, []*client.MetricMetadata{metadata1}, m.Metadata)
-	}
+				// Verify metadata
+				m, err := ing.MetricsMetadata(ctx, nil)
+				require.NoError(t, err)
+				assert.Equal(t, []*client.MetricMetadata{metadata1}, m.Metadata)
+			}
 
-	testLimits()
+			testLimits()
 
-	// Limits should hold after restart.
-	services.StopAndAwaitTerminated(context.Background(), ing) //nolint:errcheck
-	_, ing = newTestStore(t, cfg, defaultClientTestConfig(), limits, nil)
-	defer services.StopAndAwaitTerminated(context.Background(), ing) //nolint:errcheck
+			// Limits should hold after restart.
+			services.StopAndAwaitTerminated(context.Background(), ing) //nolint:errcheck
+			ing = ingGenerator()
+			defer services.StopAndAwaitTerminated(context.Background(), ing) //nolint:errcheck
+
+			testLimits()
+		})
+	}
 
-	testLimits()
 }
 
 func TestIngesterMetricLimitExceeded(t *testing.T) {
-	dirname, err := ioutil.TempDir("", "limits")
+	limits := defaultLimitsTestConfig()
+	limits.MaxLocalSeriesPerMetric = 1
+	limits.MaxLocalMetadataPerMetric = 1
+
+	dir, err := ioutil.TempDir("", "limits")
 	require.NoError(t, err)
 	defer func() {
-		require.NoError(t, os.RemoveAll(dirname))
+		require.NoError(t, os.RemoveAll(dir))
 	}()
 
-	cfg := defaultIngesterTestConfig()
-	cfg.WALConfig.WALEnabled = true
-	cfg.WALConfig.Recover = true
-	cfg.WALConfig.Dir = dirname
-	cfg.WALConfig.CheckpointDuration = 100 * time.Minute
-
-	limits := defaultLimitsTestConfig()
-	limits.MaxLocalSeriesPerMetric = 1
-	limits.MaxLocalMetadataPerMetric = 1
+	chunksDir := filepath.Join(dir, "chunks")
+	blocksDir := filepath.Join(dir, "blocks")
+	require.NoError(t, os.Mkdir(chunksDir, os.ModePerm))
+	require.NoError(t, os.Mkdir(blocksDir, os.ModePerm))
 
-	_, ing := newTestStore(t, cfg, defaultClientTestConfig(), limits, nil)
+	chunksIngesterGenerator := func() *Ingester {
+		cfg := defaultIngesterTestConfig()
+		cfg.WALConfig.WALEnabled = true
+		cfg.WALConfig.Recover = true
+		cfg.WALConfig.Dir = chunksDir
+		cfg.WALConfig.CheckpointDuration = 100 * time.Minute
 
-	userID := "1"
-	labels1 := labels.Labels{{Name: labels.MetricName, Value: "testmetric"}, {Name: "foo", Value: "bar"}}
-	sample1 := client.Sample{
-		TimestampMs: 0,
-		Value:       1,
-	}
-	sample2 := client.Sample{
-		TimestampMs: 1,
-		Value:       2,
+		_, ing := newTestStore(t, cfg, defaultClientTestConfig(), limits, nil)
+		return ing
 	}
-	labels3 := labels.Labels{{Name: labels.MetricName, Value: "testmetric"}, {Name: "foo", Value: "biz"}}
-	sample3 := client.Sample{
-		TimestampMs: 1,
-		Value:       3,
+
+	blocksIngesterGenerator := func() *Ingester {
+		ing, err := newIngesterMockWithTSDBStorageAndLimits(defaultIngesterTestConfig(), limits, blocksDir, nil)
+		require.NoError(t, err)
+		require.NoError(t, services.StartAndAwaitRunning(context.Background(), ing))
+		// Wait until it's ACTIVE
+		test.Poll(t, 10*time.Millisecond, ring.ACTIVE, func() interface{} {
```

Comment:
Let's be more generous to avoid flaky tests.

```suggestion
		test.Poll(t, time.Second, ring.ACTIVE, func() interface{} {
```